### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by amashutikov**
[DEMO LINK](https://amashutikov.github.io/react_add-todo-form/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,61 +1,123 @@
...
+    value: string;
+  }
+}
 
 export const App = () => {
+  const [userNameToAdd, setUserNameToAdd] = useState('');
+  const [titleToAdd, setTitleToAdd] = useState('');
+  const [userIsSelected, setUserIsSelected] = useState(false);
+  const [titleIsFilled, setTitleIsFilled] = useState(false);
+  const [ToDos, setToDos] = useState(todosFromServer);
```
Comment (silchyn) [inline src/App.tsx:22]:
```suggestion
  const [username, setUsername] = useState<string | null>(null);
  const [title, setTitle] = useState<string | null>(null);
  const [isUserSelected, setIsUserSelected] = useState(false);
  const [isTitleValid, setIsTitleValid] = useState(false);
  const [todos, setTodos] = useState(todosFromServer);
```

File: src/App.tsx [hunk 2]
```diff
@@ -1,61 +1,123 @@
...
+            data-cy="userSelect"
+            id="userSelect"
+          >
+            <option value="" disabled>Choose a user</option>
+
+            {usersFromServer.map(user => (
+              <option value={user.name}>
+                {user.name}
+              </option>
+            ))}
```
Comment (silchyn) [inline src/App.tsx:104]: What about the `key`?

File: src/components/UserInfo/UserInfo.tsx [hunk 3]
```diff
@@ -1 +1,14 @@
-export const UserInfo = () => {};
+import { User } from '../../types/User';
+import './UserInfo.scss';
+
+type Props = {
+  user: User | null;
```
Comment (silchyn) [inline src/components/UserInfo/UserInfo.tsx:5]: Should it accept `null`?

File: src/components/UserInfo/UserInfo.tsx [hunk 4]
```diff
@@ -1 +1,14 @@
...
+};
+
+export const UserInfo: React.FC<Props> = ({ user }) => {
```
Comment (silchyn) [inline src/components/UserInfo/UserInfo.tsx:8]: Destructure `user`

File: src/helpers.tsx [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+};
+
+export const prepareTodo = (userName: string, title: string) => {
+  const userToAddToDo = usersFromServer
+    .find(user => user.name === userName);
+  const freeId = Math.max(...todosFromServer.map(todo => todo.id)) + 1;
+
+  if (!userToAddToDo) {
+    throw new Error('No user found');
+  }
```
Comment (silchyn) [inline src/helpers.tsx:18]:
```suggestion
  const user = usersFromServer
    .find(user => user.name === userName);
    
  const id = Math.max(...todosFromServer.map(todo => todo.id)) + 1;

  if (!user) {
    throw new Error('User was not found');
  }
```

**Review by silchyn — CHANGES_REQUESTED** [review]

**Review by silchyn — CHANGES_REQUESTED** [review]
The form can be submitted with a title consisting of spaces

**Review by dinaminahi — APPROVED** [review]
Well done 👍

**Pull request closed without merging**